### ISSUE TITLE
[BUG] datetime.timedelta broken with ModelSchema

### ISSUE THREAD

**Issue by winrid** [vitalik/django-ninja#793]
Let's say I have the model:

```
class Example-(models.Model):
    when = models.DateTimeField()
    duration = models.DurationField()
```

...and the API definition:

```
from datetime import datetime, timedelta

class TestSchema(ModelSchema):
    when: datetime
    duration: timedelta
```

This breaks.
The generated OpenAPI spec will say `when` is a `string`, but `duration` is a number. However, when serializing it becomes a `string` - this breaks every generated client for statically typed languages (java, rust, etc).

using `duration: str` does not work either as then pydantic says it's not an instance of str (which is correct).

**Versions:**
 - Python version: 3.10
 - Django version: 4.2
 - Django-Ninja version: 0.22.2
 - Pydantic version: 1.10.11

**Comment by vitalik** [comment]
Hi @winrid 

well this is bascially how i[t's defined on pydantic](https://docs.pydantic.dev/1.10/usage/types/#datetime-types)

timdeltas can accept ints, floats and duration ISO format, while by default it serializes to string(ISO), you can modify it by adding [custom renderer + json encoder](https://django-ninja.rest-framework.com/guides/response/response-renderers/#create-a-renderer):

```Python

from ninja.renderers import JSONRenderer, NinjaJSONEncoder


class CustomJsonEncoder(NinjaJSONEncoder):
    def default(self, v):
        if isinstance(v, timedelta):
            return v.total_seconds() # <----- !!!!!!!
        return super().default(v)


class CustomJsonRenderer(JSONRenderer):
    encoder_class = CustomJsonEncoder


api = NinjaAPI(renderer=CustomJsonRenderer())


class TimeEventSchema(Schema):
    when: datetime
    duration: timedelta


@api.post("/event", response=TimeEventSchema)
def test_event(request, payload: TimeEventSchema):
    obj = TimeEvent.objects.create(**payload.dict())
    return obj

```

**Comment by winrid** [comment]
@vitalik  I know. But the default behavior is still broken - docs say is an int but its a str.

**Comment by winrid** [comment]
@vitalik So far, looking at Pydantic, I see that it says type is string:

```
    def timedelta_schema(self, schema: core_schema.TimedeltaSchema) -> JsonSchemaValue:
        """Generates a JSON schema that matches a timedelta value.

        Args:
            schema: The core schema.

        Returns:
            The generated JSON schema.
        """
        return {'type': 'string', 'format': 'duration'}
```

So not sure why the OpenAPI spec from Ninja says `int`. Haven't found why yet.

**Comment by winrid** [comment]
@vitalik Yes, I already have a workaround. I'm just hoping we can fix the bug. Dates work fine by default, but timedeltas do not.

**Comment by vitalik** [comment]
well this is not something fixable - as basically we just rely on the pydantic default behaviour (and changing default is a pretty breaking change)

but with pydantic2 there will be a way to add flexibility there

**Comment by winrid** [comment]
Ah okay. That's good that in pydantic2 we can fix it.

In terms of it being a breaking change, I wonder if it really is a breaking change? Because the issue is the OpenAPI spec. We can keep the default behavior in terms of functionality - accept and return strings - just fix the generated OpenAPI spec so that it is a `str` and not an `int`. So the API does not change. Just the spec is fixed.

**Comment by vitalik** [comment]
Well default pydantic2 behaviour seems to be "string":

```Python
from pydantic import BaseModel
from datetime import timedelta

class Data(BaseModel):
    t: timedelta

print(Data.schema())
```

pydantic V2:
```
'properties': {'t': {'format': 'duration', 'title': 'T', 'type': 'string'}}
```

pydantic V1:
```
'properties': {'t': {'title': 'T', 'type': 'number', 'format': 'time-delta'}}
```

**Comment by winrid** [comment]
Oh ok - does ninja just generate the spec from pydantic? Thanks for your replies!

**Comment by vitalik** [comment]
> does ninja just generate the spec from pydantic?

yes

**Comment by sasacocic** [comment]
Is there plan to update the version of pydantic to greater than 2 for ninja any time soon? Just curious.

**Comment by OtherBarry** [comment]
There are currently alpha releases which use pydantic 2. I would imagine a full release is not far off.
